### ISSUE TITLE
Check of ADCP bin mapping algorithm

### ISSUE THREAD

**Issue by BecCowley** [aodn/imos-toolbox#753]
I've been reading a lot of literature to check the bin mapping algorithm used in adcpBinMappingPP.m. These are the lines I have reviewed:

        number_of_beams = 4;
        %TODO: block function.
        CP = cos(pitch);
        % H[TxB] = P[T] x (+-R[T] x B[B])
        nonMappedHeightAboveSensorBeam1 = CP .* (cos(beamAngle + roll) / cos(beamAngle) .* distAlongBeams);
        nonMappedHeightAboveSensorBeam2 = CP .* (cos(beamAngle - roll) / cos(beamAngle) .* distAlongBeams);

        % H[TxB] = R[T] x (-+P[T] x B[B])
        CR = cos(roll);
        nonMappedHeightAboveSensorBeam3 = CR .* (cos(beamAngle - pitch) / cos(beamAngle) .* distAlongBeams);
        nonMappedHeightAboveSensorBeam4 = CR .* (cos(beamAngle + pitch) / cos(beamAngle) .* distAlongBeams);

I have two concerns.

1. The algorithm only works for up-facing ADCPs.
2. The multiplication with CP and CR is incorrect.

Here is what I think should be happening:
First, account for up or down-facing sign for pitch:

    if ~isUpwardLooking %for down facing adcps
        sg1 = 1; sg3 = 1;
    else %for up looking
        sg1 = 1; sg3 = -1;
    end

 Then, the algorithm should look like this:

        number_of_beams = 4;
        %TODO: block function.
        % H[TxB] = P[T] x (+-R[T] x B[B])
        nonMappedHeightAboveSensorBeam1 = (cos(beamAngle + sg1*roll) / cos(beamAngle) .* distAlongBeams);
        nonMappedHeightAboveSensorBeam2 = (cos(beamAngle + -sg1*roll) / cos(beamAngle) .* distAlongBeams);

        % H[TxB] = R[T] x (-+P[T] x B[B])
        nonMappedHeightAboveSensorBeam3 = (cos(beamAngle + sg3*pitch) / cos(beamAngle) .* distAlongBeams);
        nonMappedHeightAboveSensorBeam4 = (cos(beamAngle + -sg3*pitch) / cos(beamAngle) .* distAlongBeams);

I have confirmed this with the Scripps processing tools, trigonometry review, and all the literature I can find. However, when I compare this with the bin-mapped data output from the RDI Velocity processing software, the current method used in the toolbox agrees better. I'm going around in circles a bit. 

@ggalibert and @ocehugo, maybe you can explain the maths behind the algorithm currently in the toolbox? Am I missing something here? Happy to review in a call if easier.

**Comment by petejan** [comment]
@BecCowley The multiplication by CP and CR is needed otherwise this is along the beam 1-2 plane in the case of CP, not the vertical (earth) distance from the sensor. Its not clear why its / cos (beamAngle) to me.

**Comment by BecCowley** [comment]
@petejan can we catch up and go through this please?

**Comment by sspagnol** [comment]
Hi @BecCowley, can you post the links for the Scripps tools. And how does all this relate to the maths in RDI doc below?

[Appendix C - Coordinate XFORM Trig Functions.pdf](https://github.com/aodn/imos-toolbox/files/7124910/Appendix.C.-.Coordinate.XFORM.Trig.Functions.pdf)

**Comment by BecCowley** [comment]
@sspagnol thanks for the document, I've not seen this before and I really needed it! I will review.
I've put the scripps tools on cloudstor here: https://cloudstor.aarnet.edu.au/plus/s/G0AeoN4Bu98EwYk
Bin-mapping and coordinate transformation is controlled in WHbeam_ProcessFunctionVectorized.m

**Comment by BecCowley** [comment]
After much investigation into trigonometry, experiments with RDI Velocity software, Scripps software and the toolbox, and with @petejan's help, here are my very brief conclusions:

- The toolbox bin mapping algorithm is fine, except that it is only set up for down-facing instruments. Any up-facing instruments that have used this bin mapping PP tool should possibly be re-processed. I will submit a bug request to incorporate the bin-mapping for up-facing instruments.
- RDI Velocity software and the toolbox have identical results with conversion to ENU from beam velocities, when no bin-mapping is included
- When conversion to ENU is done with bin mapping, the bin mapping in the toolbox gives different results to the RDI Velocity software. I put this down to the method used (toolbox uses an interpolation, RDI uses the nearest cell). 
- The differences between the two bin-mapping methods are mostly within +/-0.2m/s in U and V (up-facing) and +/-0.1m/s (down-facing) in the test dataset I used (EAC0500, 2019-2021 data from upward 300khz and downward 75kHz instruments in single ping mode). Most of the larger offsets (up to a maximum of 0.5m/s) occurred in bins beyond the surface or bottom hits. Some occurred in bins where there is ringing or fish interference.

I can provide test datasets and figures if anyone is interested.
I couldn't code the RDI trig functions in the document that @sspagnol provided to replicate anything I got out of the Velocity software or the toolbox. But, I did have to adapt it a bit as it works from raw hertz and I have the beam velocities. Possibly I missed something....

Also, tried processing the test dataset with the Scripps functions, but again, had to hack it a bit to get it to output what I wanted with no QC filtering, and not gridded onto a standard depth grid. Possibly I didn't hack it properly....

I'm satisfied that the toolbox output matches very nearly what the RDI Velocity software is producing, once the bin mapping has been fixed to allow for up-facing instruments.
